### PR TITLE
Add #197 - 2 OutputPath (Models and Context)

### PR DESCRIPTION
**Pull request by Wasseratem**

### CODE REVIEW

**Comment by ErikEJ** [comment]
Build failed...

**Comment by Wasseratem** [comment]
Well, I see that, but I can't see why. It is working on my machine and I can't see an reason on the link at "appveyor" (what ever this tool is)

**Comment by ErikEJ** [comment]
You forgot to check in a file, I think:
CSC : error CS2001: Source file 'C:\projects\efcorepowertools\src\GUI\ReverseEngineer20\ReverseEngineer\OutputPath.cs' could not be found. [C:\projects\efcorepowertools\src\GUI\ReverseEngineer20\ReverseEngineer20.csproj]

Also, (re missing design discussion) - I propose that we stay backwards compatible, and just add a textbox below the current output path, for an optional path for the DbContext. If nothing is entered, just use the old logic, if entered, override the DbContext path.

File: src/GUI/EFCorePowerTools.Shared/Models/ModelingOptionsModel.cs [hunk 1]
```diff
@@ -17,7 +17,8 @@ public class ModelingOptionsModel : INotifyPropertyChanged
         private bool _usePluralizer;
         private bool _useDatabaseNames;
         private string _ns;
-        private string _outputPath;
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools.Shared/Models/ModelingOptionsModel.cs]: Keep this to stay backwards compatible, and just add new optional path for DbContext

File: src/GUI/EFCorePowerTools.sln [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.27130.2003
+# Visual Studio Version 16
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools.sln]: Revert this

File: src/GUI/EFCorePowerTools/Extensions/ReverseEngineerOptionsExtensions.cs [hunk 3]
```diff
@@ -24,6 +24,10 @@ public static ReverseEngineerOptions TryRead(string optionsPath)
             if (couldRead)
                 return deserialized;
 
+            couldRead = TryRead(optionsPath, out ReverseEngineerOptionsV2 deserializedV2);
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Extensions/ReverseEngineerOptionsExtensions.cs]: Stay backwards compatible

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineerHandler.cs [hunk 4]
```diff
@@ -183,7 +184,11 @@ public async void ReverseEngineerCodeFirst(Project project)
                     ContextClassName = modelingOptionsResult.Payload.ModelName,
                     DatabaseType = (ReverseEngineer20.DatabaseType)dbInfo.DatabaseType,
                     ProjectPath = projectPath,
-                    OutputPath = modelingOptionsResult.Payload.OutputPath,
+                    OutputPath = new OutputPath()
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Handlers/ReverseEngineerHandler.cs]: Just add a single new optioanl property

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineerHandler.cs [hunk 5]
```diff
@@ -271,6 +276,7 @@ public async void ReverseEngineerCodeFirst(Project project)
                     _package.LogError(revEngResult.EntityWarnings, null);
                 }
                 Telemetry.TrackEvent("PowerTools.ReverseEngineer");
+                throw new Exception("Test");
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Handlers/ReverseEngineerHandler.cs]: Remove this line

File: src/GUI/ReverseEngineer20/EFCoreReverseEngineer.cs [hunk 6]
```diff
@@ -125,7 +122,7 @@ public EfCoreReverseEngineerResult GenerateFiles(ReverseEngineerOptions reverseE
                         : reverseEngineerOptions.ConnectionString,
                     reverseEngineerOptions.Tables.Select(m => m.Name).ToArray(),
                     schemas,
-                    @namespace,
+                    modelNamespace,
                     "C#",
                     null,
```
Comment (ErikEJ) [inline src/GUI/ReverseEngineer20/EFCoreReverseEngineer.cs]: You can use this for the path to the context!

**Comment by Wasseratem** [comment]
Hi Erik,

you are correct with the throw new Exception and ofc the vs 16 lines,
But what I don't understand is the second variable.
Your argument is that we should stay backwards compatible. With the new version of the configuration I am backwards compatible. I had the idea of "just adding a second variable, like OutputContext" too - but in my opinion you should never be "scared" to rename a variable to have a better understanding for everyone on your code. With "OutputPath" and "OutputPathContext" we may ask ourself sometimes "which is which".

**Comment by ErikEJ** [comment]
The issue is that thousands of users already have saved config.json files with the existing variable name.

**Comment by Wasseratem** [comment]
This isnt an issue.
As I told you, I developed a version of the config to merge this. That is the reason I made OutputPath a Class, Not 2 new strings (and ignore the old one).
When the JSON deserializer will come to that part with OutputPath it can't serialize it, because its a string and he is now awaiting an object.

See here:

`  var couldRead = TryRead(optionsPath, out ReverseEngineerOptions deserialized);
            if (couldRead)
                return deserialized;

            couldRead = TryRead(optionsPath, out ReverseEngineerOptionsV2 deserializedV2);
            if (couldRead)
                return ReverseEngineerOptions.FromV2(deserializedV2);

            couldRead = TryRead(optionsPath, out ReverseEngineerOptionsV1 deserializedV1);
            if (couldRead)
                return ReverseEngineerOptions.FromV1(deserializedV1);`

and

`  OutputPath = new OutputPath()
                {
                    Models = v2.OutputPath,
                    Context = v2.OutputPath,
                },`

This will cause the new OutputPath.Models and OutputPath.Context to the same value as the string OutputPath.
It's completly downgradeable without problems :)

**Comment by Wasseratem** [comment]
Hi Erik,

any update on my argumentation? You want me to build it your way?

File: src/GUI/ReverseEngineer20/EFCoreReverseEngineer.cs [hunk 7]
```diff
@@ -134,30 +131,57 @@ public EfCoreReverseEngineerResult GenerateFiles(ReverseEngineerOptions reverseE
...
+            var @namespace = root;
+            if (!string.IsNullOrEmpty(path))
+            {
+                @namespace += "." + path.Replace(Path.DirectorySeparatorChar, '.').Replace(Path.AltDirectorySeparatorChar, '.');
+            }
+
+            return @namespace;
+        }
+
+        private string PostProcessContext(string contextFile, ReverseEngineerOptions options, string contextNamespace, string modelNamespace)
```
Comment (ErikEJ) [inline src/GUI/ReverseEngineer20/EFCoreReverseEngineer.cs]: No need for this, just use the DbContextPath parameter mentioned above

**Comment by ErikEJ** [comment]
I am unsure, but a big fan of as little code and as small changes as possible. My proposal will enable that, I believe. Also, there are other unresolved comments.

**Comment by Wasseratem** [comment]
Well, yes, you made your point clear :) But sadly, that is not what I think what software development should be. At the moment a dev is too scared to have changes in his code, its most likly a sign of a problem with the codebase, e.g. too less tests etc. I think we should never be afraid of making changes that will help us for the rest of our life (or the life of our software, whatever comes first). 
I see that maybe I am not the right contributor for your big project :) Feel free to accept or adopt my changes of course!

**Comment by ErikEJ** [comment]
Thanks for the effort - btw I do not see any additional tests in this PR...

**Pull request closed without merging**